cmd/src/actions_exec.go

```diff
@@ -0,0 +1,452 @@
+				}
+
+				iidFile, err := ioutil.TempFile("", "src-actions-exec-image-id")
+				if err != nil {
+					return err
+				}
+				defer os.Remove(iidFile.Name())
+
+				if *verbose {
+					log.Printf("# Building Docker container for run step %d...", i)
+				}
+				cmd := exec.CommandContext(ctx, "docker", "build", "--iidfile", iidFile.Name(), "-")
+				cmd.Stdin = strings.NewReader(run.Dockerfile)
+				verboseCmdOutput(cmd)
+				if err := cmd.Run(); err != nil {
+					return errors.Wrap(err, "build docker image")
+				}
+				if *verbose {
+					log.Printf("# Done building Docker container for run step %d.", i)
+				}
+
+				iid, err := ioutil.ReadFile(iidFile.Name())
+				if err != nil {
+					return err
+				}
+				run.Image = string(iid)
+			}
+		}
+
+		// Set digests for Docker images so we don't cache action runs in 2 different images with
+		// the same tag.
+		for _, run := range action.Run {
+			if run.Type == "docker" && run.Image != "" {
+				// getDockerImageContentDigest gets the content digest for the image. Note that this
+				// is different from the "distribution digest" (which is what you can use to specify
+				// an image to `docker run`, as in `my/image@sha256:xxx`). We need to use the
+				// content digest because the distribution digest is only computed for images that
+				// have been pulled from or pushed to a registry. See
+				// https://windsock.io/explaining-docker-image-ids/ under "A Final Twist" for a good
+				// explanation.
+				getDockerImageContentDigest := func(ctx context.Context, tag string) (string, error) {
+					const none = "<none>" // `docker images` reports this when there is no value in a cell
+
+					// TODO!(sqs): is image id the right thing to use here? it is NOT the
+					// digest. but the digest is not calculated for all images (unless they are
+					// pulled/pushed from/to a registry), see
+					// https://github.com/moby/moby/issues/32016.
+					out, err := exec.CommandContext(ctx, "docker", "image", "inspect", "--format", "{{.Id}}", "--", run.Image).CombinedOutput()
+					if err != nil {
+						return "", fmt.Errorf("error inspecting docker image (try `docker pull %q` to fix this): %s", run.Image, bytes.TrimSpace(out))
+					}
+					id := string(bytes.TrimSpace(out))
+					if id == "" {
+						return "", fmt.Errorf("unexpected empty docker image content ID for %q", run.Image)
+					}
+					return id, nil
+				}
+
+				run.ImageContentDigest, err = getDockerImageContentDigest(ctx, run.Image)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		uilive.Out = os.Stderr
+		uilive.RefreshInterval = 10 * time.Hour // TODO!(sqs): manually flush
+		color.NoColor = false                   // force color even when in a pipe
+		var (
+			lwMu sync.Mutex
+			lw   = uilive.New()
+		)
+		lw.Start()
+		defer lw.Stop()
+
+		spinner := []rune{'⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'}
+		spinnerI := 0
+		onUpdate := func(reposMap map[ActionRepo]ActionRepoStatus) {
```

Comment:
Extract this and the printing into its own struct (`ProgressPrinter`?) and pass only small function as callback to executor.